executor/windowfuncs/window_funcs.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package windowfuncs
+
+import (
+	"errors"
+
+	"github.com/pingcap/tidb/executor/aggfuncs"
+	"github.com/pingcap/tidb/expression/aggregation"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+// WindowFunc is the interface for processing window functions.
+type WindowFunc interface {
+	// ProcessOneChunk processes one chunk.
+	ProcessOneChunk(sctx sessionctx.Context, rows []chunk.Row, dest *chunk.Chunk) ([]chunk.Row, error)
+	// ExhaustResult exhausts result to the result chunk.
+	ExhaustResult(sctx sessionctx.Context, rows []chunk.Row, dest *chunk.Chunk) ([]chunk.Row, error)
+	// HasRemainingResults checks if there are some remained results to be exhausted.
+	HasRemainingResults() bool
+}
+
+// aggWithoutFrame deals with agg functions with no frame specification.
+type aggWithoutFrame struct {
```

Comment:
It's better to not store the function state inside this function. Let the caller handle and store the state like aggregate functions. So that one day, we can simply control the memory usage in the caller side, no need to worry about the concrete functions.

Can we split this struct into two parts, for example:
```go
type winFuncNoFrame struct {
	...
}

type partialResult4WinFuncNoFrame struct {
	...
}
```

Reply:
Done.